### PR TITLE
chore: move new style navigation into Core/Navigation

### PR DESCRIPTION
**Pull request by asalzburger**
This PR moves the new navigation classes for the still `Experimental` geometry into a separate `Navigation` folder in order to better separate `Geometry` and `Navigation` objects.

The following classes, files and definitions build the new-style navigation approach:

* `NavigationDelegates.hpp` this files defines the different navigation structs/functions following the `Delegate=` implementation (or `OwnedDelegate` if needed)

* `NavigationState` a struct that holds navigation candidates (a `NavigationCandidate` is either a `Surface` or a `Portal`, this should be the state of the of the new `Experimental::Navigator` or at least attached to it

* `NavigationStateUpdators` these are updaters that are attached to `DetectorVolume` and `PortalObjects` that allow to update the `NavigationCandidate` list during propagation

* `DetectorVolumeFinders` are delegates that allow to find an associated `DetectorVolume` purely from position information, these can exist as brute-force (tryAll), grid (Indexed) or future boundary box implementations

* `SurfaceCandidateUpdators` are updaters that fill the surfaces within a volume into the `NavigationState`, they are pluggable components into the `NavigationStateUpdators`

### CODE REVIEW

File: Core/include/Acts/Geometry/Extent.hpp [hunk 1]
```diff
@@ -174,6 +174,11 @@ class Extent {
     return 0.5 * (m_range[bValue].min() + m_range[bValue].max());
   }
 
+  /// Access the maximum parameter
```
Comment (andiwand) [inline Core/include/Acts/Geometry/Extent.hpp:177]:
```suggestion
  /// Access the span
```

File: Core/include/Acts/Geometry/ProtoDetector.hpp [hunk 2]
```diff
@@ -13,12 +13,32 @@
...
+
+namespace Experimental {
+class DetectorVolume;
+class Portal;
+
+/// Current volumes (connected)
+using DetectorVolumes = std::vector<std::shared_ptr<DetectorVolume>>;
+/// Current skin (outside volumes)
+using ProtoContainer = std::map<unsigned int, std::shared_ptr<Portal>>;
+/// Current blok (volumes and skin)
```
Comment (andiwand) [inline Core/include/Acts/Geometry/ProtoDetector.hpp:34]:
```suggestion
/// Current block (volumes and skin)
```

**Review by andiwand — COMMENTED** [review]
had a quick look - will dive in in a second round

File: Core/include/Acts/Geometry/Extent.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/Extent.hpp]: Fixed.

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 2 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp]: added c.

File: Core/include/Acts/Geometry/Extent.hpp [hunk 3]
```diff
@@ -174,6 +174,11 @@ class Extent {
     return 0.5 * (m_range[bValue].min() + m_range[bValue].max());
   }
 
+  /// Access the parameter span
+  ///
+  /// @param bValue the binning identification
+  ActsScalar span(BinningValue bValue) const { return m_range[bValue].size(); }
```
Comment (paulgessinger) [inline Core/include/Acts/Geometry/Extent.hpp:180]: Is `span` a good name for this?

File: Core/include/Acts/Geometry/ProtoDetector.hpp [hunk 4]
```diff
@@ -13,12 +13,32 @@
...
 
+struct ProtoVolume;
+
+namespace Experimental {
+class DetectorVolume;
+class Portal;
+
+/// Current volumes (connected)
+using DetectorVolumes = std::vector<std::shared_ptr<DetectorVolume>>;
+/// Current skin (outside volumes)
```
Comment (paulgessinger) [inline Core/include/Acts/Geometry/ProtoDetector.hpp:32]: skin?

File: Core/include/Acts/Geometry/detail/PortalHelper.hpp [hunk 5]
```diff
@@ -73,6 +73,60 @@ defaultPortalAndSubPortalGenerator() {
   return pGenerator;
 }
 
+/// Definition of a portal replacement when building proto containers
+/// It consists of the new portal, the index, the direction, the parameters
+/// gathered from the sub volumes, the binning description
+using PortalReplacement =
+    std::tuple<std::shared_ptr<Experimental::Portal>, unsigned int,
+               NavigationDirection, std::vector<ActsScalar>, BinningValue>;
```
Comment (paulgessinger) [inline Core/include/Acts/Geometry/detail/PortalHelper.hpp:81]: Could this be an extra `struct` with names? This has 5 elements and might start being confusing somewhat.

Note you can still destructure this like a tuple if it's plain struct like you do below, I believe.

```cpp
for (auto& [p, i, dir, boundaries, binning] : pReplacements) {}
```

File: Core/include/Acts/Geometry/detail/PortalHelper.hpp [hunk 6]
```diff
@@ -73,6 +73,60 @@ defaultPortalAndSubPortalGenerator() {
...
+std::vector<std::shared_ptr<DetectorVolume>> attachedDetectorVolumes(
+    Portal& portal) noexcept(false) {
+  auto& attachedVolumes = portal.attachedDetectorVolumes();
+  if (not attachedVolumes[0u].empty() and not attachedVolumes[1u].empty()) {
+    throw std::invalid_argument(
+        "PortalHelper: trying to get attachedVolumes from already populated "
+        "portal.");
+  }
+  unsigned int iu = attachedVolumes[0u].empty() ? 1u : 0u;
+  return attachedVolumes[iu];
```
Comment (paulgessinger) [inline Core/include/Acts/Geometry/detail/PortalHelper.hpp:127]: We're not perf critical here, so maybe

```suggestion
  return attachedVolumes.at(iu);
```

File: Core/include/Acts/Geometry/detail/PortalHelper.hpp [hunk 7]
```diff
@@ -73,6 +73,60 @@ defaultPortalAndSubPortalGenerator() {
...
+/// @brief  Method that strips out attached volumes from portals
+///
+/// @note it throws an exception if both sides are already taken
+///
+/// @param portal the portal to be resolved
+///
+/// @return a vector of attached volumes
+std::vector<std::shared_ptr<DetectorVolume>> attachedDetectorVolumes(
+    Portal& portal) noexcept(false) {
+  auto& attachedVolumes = portal.attachedDetectorVolumes();
```
Comment (paulgessinger) [inline Core/include/Acts/Geometry/detail/PortalHelper.hpp:120]:
```suggestion
  const auto& attachedVolumes = portal.attachedDetectorVolumes();
```

File: Core/include/Acts/Geometry/Extent.hpp (hunk 3 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/Extent.hpp:180]: does it not 'span' from min to max ?

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 4 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:32]: 'Shell' ?

File: Core/include/Acts/Geometry/detail/PortalHelper.hpp (hunk 6 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/detail/PortalHelper.hpp:127]: Sounds good.

File: Core/include/Acts/Geometry/detail/PortalHelper.hpp (hunk 5 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/detail/PortalHelper.hpp:81]: Ok, will look into this, if i'ts deconstructable it's fine.

File: Core/include/Acts/Geometry/detail/PortalHelper.hpp (hunk 7 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/detail/PortalHelper.hpp:120]: Understood.

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 4 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:32]: I changed the description.

File: Core/include/Acts/Geometry/Extent.hpp (hunk 3 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/Extent.hpp:180]: changed to `interval()`

File: Core/include/Acts/Geometry/detail/PortalHelper.hpp (hunk 5 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/detail/PortalHelper.hpp:81]: Will move this comment to the `Detector` PR where these classes are touched again, and will try it.

**Comment by asalzburger** [comment]
Will add an automerge, since this doesn't change a line of code.

**Review by andiwand — APPROVED** [review]

**Comment by andiwand** [comment]
makes sense I am going to study the code further offline

**Pull request merged**